content/blog/2020-11-29-tracemonkey/index.md

```diff
@@ -0,0 +1,402 @@
+part by referencing the oracle and not speculatively assigning any
+variable to be an integer if the oracle says otherwise.
+
+### Optimizations
+
+After the trace is recorded as LIR, nanojit runs a simple set of
+optimization passes followed by register allocation to produce machine
+code. As the LIR is a linear sequence of instructions in SSA form, all
+optimizations can be done in linear time and are straightforwardly
+local.
+
+The optimization passes are noticeably simpler than those used by
+static compilers because it is important not to spend too much time
+compiling while the program is executing. Other more recent JavaScript
+engines, including V8, go back and re-compile the same code with more
+sophisticated optimizations if it stays hot. TraceMonkey does not do
+this; perhaps such things are future improvements, or perhaps it was
+preferred to focus on the optimizations that deliver the best bang for
+your buck.
+
+In addition to the speculative integer conversion, the following
+optimizations are performed by nanojit:
+
+ - Common subexpression elimination, a standard optimization (forward
+   analysis)
+   
+ - Constant folding, another standard optimization (forward analysis)
+
+ - Some pinhole optimizations for identities described by the authors,
+   e.g. `0 + x` being reduced to just `x` (forward analysis)
+
+ - Dead data-stack store elimination, essentially removing unused
+   stack-allocated variables (backward analysis)
+
+ - Dead call-stack store elimination, essentially removing unused
+   stack frames code as a result of function inlining (backward
+   analysis)
+
+ - Dead code elimination, a standard optimization (forward analysis)
+
+The optimizations run in one forward pass and one backward pass, which
+includes greedy register allocation.
+
+Also worth mentioning is that tracing implicitly grants function
+inlining for free because all traces follow function calls into
+subroutines, so all of the above optimizations have the power of
+interprocedural optimizations with the simplicity of local SSA
+optimizations.
+
+### Trace stitching
+
+Just as a compiled trace can loop back to its own header, tying the
+knot, to make the compiled code more efficient, it is beneficial to be
+able to directly jump to a branch trace in the event of a guard
+failure. It is in general difficult to perform this jump directly
+without returning to the monitor because the register allocation was
+performed separately.
+
+Enter branch stitching. The idea is to recompile an entire tree
+together when new branches are added, which allows the branches to be
+simple jumps with excellent performance. The problem is that
+recompiling the entire tree is expensive, especially for trace trees
+with lots of short branches. The authors discuss this problem and
+propose performing the recompilation in parallel, a technique which
+appears to have been adopted by modern JavaScript engines today.
```

Comment:
Interesting—do you have evidence/citations for the claim that parallel codegen exists in contemporary JITs?